src/lowering/desugar.ml

```diff
@@ -68,6 +68,20 @@ and exp' at note = function
     I.PrimE (I.RelPrim (!ot, o), [exp e1; exp e2])
   | S.ShowE (ot, e) ->
     I.PrimE (I.ShowPrim !ot, [exp e])
+  | S.ToCandidE (ot, es) ->
+    begin match !ot with
+    | T.Tup ts1 ->
+      I.PrimE (I.SerializePrim [T.Tup ts1], [tupE (exps es)])
+    | _ -> assert false
+    end
+  | S.FromCandidE (ot, e) ->
+    begin match !ot with
```

Comment:
You almost certainly need to normalize !ot (or just the outer type of the entire expression) before testing the outer structure of the type, and then normalize the content again: